tests/v2/waku_filter_v2/test_waku_filter.nim

```diff
@@ -0,0 +1,287 @@
+      messagePushHandler: MessagePushHandler = proc(pubsubTopic: PubsubTopic, message: WakuMessage) =
+        pushHandlerFuture.complete((pubsubTopic, message))
+
+    let
+      serverSwitch = newStandardSwitch()
+      clientSwitch = newStandardSwitch()
+      wakuFilter = await newTestWakuFilter(serverSwitch)
+      wakuFilterClient = await newTestWakuFilterClient(clientSwitch, messagePushHandler)
+      clientPeerId = clientSwitch.peerInfo.peerId
+      pubsubTopic = DefaultPubsubTopic
+      contentTopic2 = ContentTopic("/waku/2/non-default-content/proto")
+      contentTopics = @[DefaultContentTopic, contentTopic2]
+
+    # When
+    await allFutures(serverSwitch.start(), clientSwitch.start())
+    let response = await wakuFilterClient.subscribe(serverSwitch.peerInfo.toRemotePeerInfo(), pubsubTopic, contentTopics)
+
+    # Then
+    check:
+      response.isOk()
+      wakuFilter.subscriptions.len == 1
+      wakuFilter.subscriptions.hasKey(clientPeerId)
+
+    # When
+    let msg1 = fakeWakuMessage(contentTopic=DefaultContentTopic)
+    await wakuFilter.handleMessage(DefaultPubsubTopic, msg1)
+
+    require await pushHandlerFuture.withTimeout(3.seconds)
+
+    # Then
+    let (pushedMsgPubsubTopic, pushedMsg) = pushHandlerFuture.read()
+    check:
+      pushedMsgPubsubTopic == DefaultPubsubTopic
+      pushedMsg == msg1
+
+    # When
+    let msg2 = fakeWakuMessage(contentTopic=contentTopic2)
+    pushHandlerFuture = newFuture[(string, WakuMessage)]() # Clear previous future
+    await wakuFilter.handleMessage(DefaultPubsubTopic, msg2)
+
+    require await pushHandlerFuture.withTimeout(3.seconds)
+
+    # Then
+    let (pushedMsgPubsubTopic2, pushedMsg2) = pushHandlerFuture.read()
+    check:
+      pushedMsgPubsubTopic2 == DefaultPubsubTopic
+      pushedMsg2 == msg2
+
+    # When
+    let response2 = await wakuFilterClient.unsubscribe(serverSwitch.peerInfo.toRemotePeerInfo(), pubsubTopic, @[contentTopic2]) # Unsubscribe only one content topic
+
+    # Then
+    check:
+      response2.isOk()
+      wakuFilter.subscriptions.len == 1
+
+    # When
+    let msg3 = fakeWakuMessage(contentTopic=DefaultContentTopic)
+    pushHandlerFuture = newFuture[(string, WakuMessage)]() # Clear previous future
+    await wakuFilter.handleMessage(DefaultPubsubTopic, msg3)
+
+    require await pushHandlerFuture.withTimeout(3.seconds)
+
+    # Then
+    let (pushedMsgPubsubTopic3, pushedMsg3) = pushHandlerFuture.read()
+    check:
+      pushedMsgPubsubTopic3 == DefaultPubsubTopic
+      pushedMsg3 == msg3
+
+    # When
+    let msg4 = fakeWakuMessage(contentTopic=contentTopic2)
+    pushHandlerFuture = newFuture[(string, WakuMessage)]() # Clear previous future
+    await wakuFilter.handleMessage(DefaultPubsubTopic, msg4)
+
+    # Then
+    check:
+      not (await pushHandlerFuture.withTimeout(2.seconds)) # No message should be pushed
+
+  asyncTest "subscribe to multiple content topics and unsubscribe all":
```

Comment:
Perhaps missing a testcase subscribing to multiple gossipsub topics and multiple content topics?

Reply:
Good point! Added in [9105f88](https://github.com/waku-org/nwaku/pull/1600/commits/9105f88a45688599ea8ee580a873aa17a3af21ac) (I want to significantly extend test cases, but at least cover the most basic happy paths for now, so this was indeed an oversight).